tests/api_client/test_api_client.py

```diff
@@ -0,0 +1,309 @@
+
+
+class MockPOSTResponse(MockResponse):
+    possible_responses = {
+        201: SAMPLE_JOB_CREATE_RESPONSE,
+        400: {
+            "code": "parse-error",
+            "detail": (
+                "The blackbird script could not be parsed. "
+                "Please fix errors in the script and try again.")
+        },
+        401: {
+            "code": "unauthenticated",
+            "detail": "Requires authentication"
+        },
+        409: {
+            "code": "unsupported-circuit",
+            "detail": (
+                "This circuit is not compatible with the specified hardware.")
+        },
+        500: {
+            "code": "server-error",
+            "detail": (
+                "Unexpected server error. Please try your request again "
+                "later.")
+        },
+    }
+
+
+class MockGETResponse(MockResponse):
+    possible_responses = {
+        200: SAMPLE_JOB_RESPONSE,
+        401: {
+            "code": "unauthenticated",
+            "detail": "Requires authentication"
+        },
+        404: {
+            "code": "not-found",
+            "detail": "The requested resource could not be found or does not exist.",
+        },
+        500: {
+            "code": "server-error",
+            "detail": (
+                "Unexpected server error. Please try your request again "
+                "later.")
+        },
+    }
+
+    status_code = None
+
+    def __init__(self, status_code):
+        self.status_code = status_code
+
+    def json(self):
+        return self.possible_responses[self.status_code]
+
+    def raise_for_status(self):
+        raise requests.exceptions.HTTPError()
+
+
+@pytest.mark.api_client
+class TestAPIClient:
+    def test_init_default_client(self):
+        client = api_client.APIClient()
+        assert client.USE_SSL is True
+        assert client.AUTHENTICATION_TOKEN is None
+        assert client.BASE_URL == 'https://localhost'
+        assert client.HEADERS == {}
+
+    def test_init_default_client_no_ssl(self):
+        client = api_client.APIClient(use_ssl=False)
+        assert client.USE_SSL is False
+        assert client.AUTHENTICATION_TOKEN is None
+        assert client.BASE_URL == 'http://localhost'
+        assert client.HEADERS == {}
+
+    def test_init_custom_token_client(self):
+        test_token = 'TEST'
+        client = api_client.APIClient(authentication_token=test_token)
+        assert client.AUTHENTICATION_TOKEN == test_token
+
+    def test_load_configuration(self, client):
+        with pytest.raises(NotImplementedError):
+            client.load_configuration_from_file()
+
+    def test_authenticate(self, client):
+        with pytest.raises(NotImplementedError):
+            username = 'TEST_USER'
+            password = 'TEST_PASSWORD'
+            client.authenticate(username, password)
+
+    def test_join_path(self, client):
+        assert client.join_path('jobs') == f'{client.BASE_URL}/jobs'
```

Comment:
Don't forget to remove all f-strings for now (we can evaluate later whether to shift the minimum supported Python version to 3.6)

Reply:
Done.